### PR TITLE
Fix VerifiableCredentialsJwtServiceImpl's "NotSuchMethodError"

### PR DESCRIPTION
**Pull request by marcgs**
## What this PR changes/adds

Introduce a fix for the "NoSuchMethodError" observed in the CI pipeline for the cloud system tests: https://github.com/agera-edc/MinimumViableDataspace/runs/7693917471?check_suite_focus=true

The cause of the issue is still unclear. It seems that the `toJSONObject` method is not available due to a version mismatch of the nimbus library. This does not happen when running tests locally, nor when running the local system tests in CI. Why this only manifests when running the cloud system tests in CI is still a mystery.

MicrosoftTeams-image

This PR introduces a workaround by adapting the `VerifiableCredentialsJwtServiceImpl` not to use this method.

Logging was improved in order to debug the issue and this PR contains additional logging statements that proved to be useful.

## Why it does that

Fix Identity Hub integration in MVD.

## Linked Issue(s)

Relates to #21

## Checklist

- [x] added appropriate tests?
- [x] performed checkstyle check locally?
- [x] added/updated copyright headers?
- [x] documented public classes/methods?
- [x] added/updated relevant documentation?
- [ ] added relevant details to the changelog? (_skip with label `no-changelog`_)
- [ ] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-dataspaceconnector/identityservice/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [styleguide](https://github.com/eclipse-dataspaceconnector/identityservice/blob/main/styleguide.md) for details_)

### CODE REVIEW

File: extensions/identity-hub-verifier/src/test/java/org/eclipse/dataspaceconnector/identityhub/verifier/CredentialsVerifierExtensionTest.java [hunk 1]
```diff
@@ -42,32 +41,28 @@
...
     void setUp(EdcExtension extension) {
         identityHubClient = new IdentityHubClientImpl(TestUtils.testOkHttpClient(), new ObjectMapper(), new ConsoleMonitor());
-        publicKeyResolver = mock(DidPublicKeyResolverImpl.class);
-        extension.registerServiceMock(DidPublicKeyResolver.class, publicKeyResolver);
         extension.setConfiguration(Map.of("web.http.port", String.valueOf(PORT)));
     }
 
     @Test
-    public void getVerifiedClaims_getValidClaims(CredentialsVerifier verifier) {
+    public void getVerifiedClaims_getValidClaims(CredentialsVerifier verifier, DidPublicKeyResolver mockResolver) {
```
Comment (ouphi) [inline extensions/identity-hub-verifier/src/test/java/org/eclipse/dataspaceconnector/identityhub/verifier/CredentialsVerifierExtensionTest.java]: Reminder to revert the test fix from this PR.

**Review by ouphi — APPROVED** [review]

File: client-cli/src/main/java/org/eclipse/dataspaceconnector/identityhub/cli/ListVerifiableCredentialsCommand.java [hunk 2]
```diff
@@ -44,6 +44,9 @@ class ListVerifiableCredentialsCommand implements Callable<Integer> {
     public Integer call() throws Exception {
         var out = spec.commandLine().getOut();
         var result = command.cli.identityHubClient.getVerifiableCredentials(command.cli.hubUrl);
+        if (result.failed()) {
+            throw new CliException("Failed to get verifiable credentials: " + String.join(",", result.getFailureMessages()));
```
Comment (algattik) [inline client-cli/src/main/java/org/eclipse/dataspaceconnector/identityhub/cli/ListVerifiableCredentialsCommand.java]:
```suggestion
            throw new CliException("Failed to get verifiable credentials: " + result.getFailureDetail());
```

File: extensions/identity-hub-verifier/src/test/java/org/eclipse/dataspaceconnector/identityhub/verifier/CredentialsVerifierExtensionTest.java [hunk 3]
```diff
@@ -50,24 +52,28 @@ public class CredentialsVerifierExtensionTest {
...
-    private static final String CREDENTIAL_ISSUER = FAKER.internet().url();
-    private static final String SUBJECT = FAKER.internet().url();
+    private static final String CREDENTIAL_ISSUER = "did:web:" + FAKER.internet().domainName();
+    private static final String SUBJECT = "did:web:" + FAKER.internet().domainName();
     private IdentityHubClient identityHubClient;
-    private DidPublicKeyResolver publicKeyResolver;
 
     @BeforeEach
     void setUp(EdcExtension extension) {
         identityHubClient = new IdentityHubClientImpl(TestUtils.testOkHttpClient(), new ObjectMapper(), new ConsoleMonitor());
```
Comment (algattik) [inline extensions/identity-hub-verifier/src/test/java/org/eclipse/dataspaceconnector/identityhub/verifier/CredentialsVerifierExtensionTest.java]: use mock console

File: spi/identity-hub-spi/src/test/java/org/eclipse/dataspaceconnector/identityhub/credentials/VerifiableCredentialsJwtServiceTest.java [hunk 4]
```diff
@@ -50,7 +51,7 @@ public void setUp() {
         var key = generateEcKey();
         privateKey = new EcPrivateKeyWrapper(key);
         publicKey = new EcPublicKeyWrapper(key);
-        service = new VerifiableCredentialsJwtServiceImpl(OBJECT_MAPPER);
+        service = new VerifiableCredentialsJwtServiceImpl(OBJECT_MAPPER, new ConsoleMonitor());
```
Comment (algattik) [inline spi/identity-hub-spi/src/test/java/org/eclipse/dataspaceconnector/identityhub/credentials/VerifiableCredentialsJwtServiceTest.java]: mock

File: extensions/identity-hub-verifier/src/main/java/org/eclipse/dataspaceconnector/identityhub/verifier/IdentityHubCredentialsVerifier.java [hunk 5]
```diff
@@ -61,41 +68,74 @@ public IdentityHubCredentialsVerifier(IdentityHubClient identityHubClient, Monit
      */
     @Override
     public Result<Map<String, Object>> getVerifiedCredentials(DidDocument didDocument) {
+        monitor.debug("Retrieving verified credentials for " + didDocument.getId());
```
Comment (algattik) [inline extensions/identity-hub-verifier/src/main/java/org/eclipse/dataspaceconnector/identityhub/verifier/IdentityHubCredentialsVerifier.java]:
```suggestion
        monitor.debug(() -> "Retrieving verified credentials for " + didDocument.getId());
```

also in other places

File: extensions/identity-hub-verifier/src/main/java/org/eclipse/dataspaceconnector/identityhub/verifier/IdentityHubCredentialsVerifier.java [hunk 6]
```diff
@@ -61,41 +68,74 @@ public IdentityHubCredentialsVerifier(IdentityHubClient identityHubClient, Monit
...
+        if (hubBaseUrl.isEmpty()) {
+            var errorMessage = "Could not retrieve identity hub URL from DID document";
+            monitor.severe(errorMessage);
+            return Result.failure(errorMessage);
         }
 
-        var jwts = identityHubClient.getVerifiableCredentials(hubBaseUrl.getContent());
+        monitor.debug(String.format("Using identity hub URL: %s", hubBaseUrl.get()));
+
+        var jwts = identityHubClient.getVerifiableCredentials(hubBaseUrl.get());
```
Comment (algattik) [inline extensions/identity-hub-verifier/src/main/java/org/eclipse/dataspaceconnector/identityhub/verifier/IdentityHubCredentialsVerifier.java]: use a better name than `jwts`?

File: extensions/identity-hub-verifier/src/main/java/org/eclipse/dataspaceconnector/identityhub/verifier/IdentityHubCredentialsVerifier.java [hunk 7]
```diff
@@ -61,41 +68,74 @@ public IdentityHubCredentialsVerifier(IdentityHubClient identityHubClient, Monit
...
+        var result = verifiedCredentials.stream()
+                .map(verifiableCredentialsJwtService::extractCredential)
+                .collect(partitioningBy(AbstractResult::succeeded));
+
+        var successfulResults = result.get(true);
+        var failedResults = result.get(false);
 
-        var claims = successfulResults.stream()
+        if (!failedResults.isEmpty()) {
+            failedResults.forEach(f -> monitor.warning("Invalid credentials: " + String.join(",", f.getFailureMessages())));
```
Comment (algattik) [inline extensions/identity-hub-verifier/src/main/java/org/eclipse/dataspaceconnector/identityhub/verifier/IdentityHubCredentialsVerifier.java]: use `getFailureDetail`

File: extensions/identity-hub-verifier/src/test/java/org/eclipse/dataspaceconnector/identityhub/verifier/CredentialsVerifierExtensionTest.java [hunk 8]
```diff
@@ -50,24 +52,28 @@ public class CredentialsVerifierExtensionTest {
...
-    public void getVerifiedClaims_getValidClaims(CredentialsVerifier verifier) {
-        // Arrange
+    public void getVerifiedClaims_getValidClaims(CredentialsVerifier verifier, DidResolverRegistry registry) {
+
         var jwk = generateEcKey();
-        when(publicKeyResolver.resolvePublicKey(anyString())).thenReturn(Result.success(toPublicKeyWrapper(jwk)));
+        // Arrange - add did resolver that returns a dummy DID
+        var didResolver = mock(DidResolver.class);
+        var did = createDidDocument(jwk);
+        when(didResolver.resolve(anyString())).thenReturn(Result.success(did));
```
Comment (algattik) [inline extensions/identity-hub-verifier/src/test/java/org/eclipse/dataspaceconnector/identityhub/verifier/CredentialsVerifierExtensionTest.java]: nit: can we be more specific than `anyString`

**Review by algattik — CHANGES_REQUESTED** [review]

**Review by algattik — APPROVED** [review]

**Pull request merged**